src/System.Text.Json/System/Text/Json/JsonWriter.cs

```diff
@@ -6,193 +6,592 @@
+
+        /// <summary>
+        /// Write a <see cref="Guid"/> value into the current array.
+        /// </summary>
+        /// <param name="value">The <see cref="Guid"/> value to be written to JSON data.</param>
+        public void WriteValue(Guid value)
         {
-            ConsideComma();
-            WriteIndent();
-            _formatter.Append('"');
-            _formatter.Append(name);
-            _formatter.Append("\":");
+            WriteItemSeperator();
+            _firstItem = false;
+            WriteSpacing();
+            WriteGuid(value);
         }
 
-        private void ConsideComma()
+        /// <summary>
+        /// Write a null value into the current array.
+        /// </summary>
+        public void WriteNull()
         {
-            if (_wroteListItem) {
-                _formatter.Append(',');
-                WriteNewline();
-            }
+            WriteItemSeperator();
+            _firstItem = false;
+            WriteSpacing();
+            WriteAscii(JsonConstants.NullValue);
+        }
+
+        /// <summary>
+        /// Write an undefined value into the current array.
+        /// </summary>
+        public void WriteUndefined()
+        {
+            WriteItemSeperator();
+            _firstItem = false;
+            WriteSpacing();
+            WriteAscii(JsonConstants.UndefinedValue);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteStartAttribute(string name)
+        {
+            WriteItemSeperator();
+            _firstItem = false;
+
+            WriteSpacing();
+            WriteQuotedString(name);
+            WriteChar(JsonConstants.KeyValueSeperator);
+
+            if (_prettyPrint)
+                WriteChar(JsonConstants.Space);
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private void WriteNewline()
+        private void WriteChar(char value)
         {
-            if (PrettyPrint) {
-                _formatter.Append('\n');
+            var output = _output;
+            if (UseFastUtf8)
+            {
+                if (output.Buffer.Length < 1)
+                    output.Enlarge();
+
+                output.Buffer.DangerousGetPinnableReference() = (byte)value;
+                output.Advance(1);
             }
+            else if (UseFastUtf16)
+            {
+                if (output.Buffer.Length < 2)
+                    output.Enlarge(2);
+
+                unsafe
+                {
+                    Unsafe.AsRef<char>(Unsafe.AsPointer<byte>(ref output.Buffer.DangerousGetPinnableReference())) = value;
+                }
+
+                output.Advance(2);
+            }
+            else
+            {
+                unsafe
+                {
+                    // Slow path, if we are dealing with non-invariant.
+                    ReadOnlySpan<char> chSpan = new ReadOnlySpan<char>(&value, 1);
+                    Write(chSpan);
```

Comment:
It would be good to add per benchmarks for non-invariant case. It's going to be slower, but we cannot be ridiculously slow.